devito/mpi/distributed.py

```diff
@@ -421,6 +440,197 @@ def _rebuild(self, shape=None, dimensions=None, comm=None):
+
+        self._dimension_map = frozendict({pd: sd for pd, sd
+                                          in zip(subdomain.grid.dimensions,
+                                                 subdomain.dimensions)})
+
+        self._parent = subdomain.grid.distributor
+
+        self._comm = self.parent.comm
+        self._topology = self.parent.topology
+
+        # Want to calculate and keep track of crosses here
+        self.__align_parent__()
+
+    def __align_parent__(self):
+        """
+        Routines for determining alignment of the SubDomain which this
+        SubDomainDistributor decomposes relative to the overarching
+        decomposition of the grid.
+        """
+        dist_interval = {dim: Interval(s.start, s.stop-1)
+                         for dim, s in self.parent.glb_slices.items()}
+
+        # Assumes no override of x_m and x_M supplied to operator
+        bounds_map = {**{dim.symbolic_min: 0
+                         for dim in self.parent.dimensions},
+                      **{dim.symbolic_max: sha-1
+                         for dim, sha in zip(self.parent.dimensions,
+                                             self.parent.glb_shape)}}
+
+        sdim_interval = {dim:
+                         (sdim._interval.subs({**{k: v for k, v
+                                                  in sdim._thickness_map.items()
+                                                  if v is not None},
+                                               **bounds_map})
+                          if sdim.is_Sub else None)
+                         for dim, sdim in zip(self.parent.dimensions, self.dimensions)}
+
+        # The domain decomposition
+        decompositions = []
+        for dec, dim in zip(self.parent._decomposition, self.parent.dimensions):
+            decompositions.append([d if sdim_interval[dim] is None
+                                   else np.zeros(0, dtype=np.int64)
+                                   if sdim_interval[dim].is_empty
+                                   else d[np.logical_and(d >= sdim_interval[dim].start,
+                                                         d <= sdim_interval[dim].end)]
+                                   if sdim_interval[dim].is_Interval
+                                   else np.array(sdim_interval[dim].args, dtype=np.int64)
+                                   for d in dec])
+
+        self._decomposition = [Decomposition(d, c)
+                               for d, c in zip(decompositions, self.mycoords)]
+
+        self._intervals = tuple((dist_interval[dim] if sdim_interval[dim] is None
+                                 else dist_interval[dim].intersect(sdim_interval[dim]))
+                                for dim in self.parent.dimensions)
+
+        # Keep track of the edges of the rank crossed by the SubDomain
+        # Format is {dimension: {side: crosses}} or {(sides,): crosses}
+        # (matches neighborhood)
+        crosses = {}
```

Comment:
do these things really need to be precomputed at `__init__`? can we not move them into `cached_property`'s ? I'm talking about the `crosses` here, but also the `decomposition`, `intervals`, etc above

Reply:
Did you think about introducing `SubDecomposition`'s or similar? With such ideas and some nice little tools (maps) maybe a lot of the verbosity above can be removed?